test/e2e/metrics_test.go

```diff
@@ -89,134 +129,265 @@ func NewMetricsTestConfig(t *testing.T, client, selector, clusterRole, clusterBi
+func (c *MetricsTestConfig) waitForPodReady(ctx context.Context) {
 	c.t.Log("Waiting for the curl pod to be ready")
-	waitCmd := exec.Command(c.client, "wait", "--for=condition=Ready", "pod", c.curlPodName, "-n", c.namespace, "--timeout=60s")
-	waitOutput, waitErr := waitCmd.CombinedOutput()
-	require.NoError(c.t, waitErr, "Error waiting for curl pod to be ready: %s", string(waitOutput))
+	err := wait.PollUntilContextTimeout(ctx, 2*time.Second, 60*time.Second, false, func(ctx context.Context) (bool, error) {
+		pod, err := c.kubeClient.CoreV1().Pods(c.namespace).Get(ctx, c.curlPodName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		for _, cond := range pod.Status.Conditions {
+			if cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if err != nil {
+		// If the context timed out, the test should fail with a more direct message
+		if errors.Is(err, context.DeadlineExceeded) {
+			c.t.Fatal("Timed out waiting for the curl pod to become Ready")
+		}
+		require.NoError(c.t, err, "Error waiting for curl pod to become Ready")
+	}
+}
+
+// validateMetricsEndpoint performs `kubectl exec ... curl <metricsURL>` logic
+func (c *MetricsTestConfig) validateMetricsEndpoint(ctx context.Context, token string) {
+	c.t.Log("Validating the metrics endpoint via pod exec")
+
+	// The command to run inside the container
+	cmd := []string{
+		"curl", "-v", "-k",
+		"-H", "Authorization: Bearer " + token,
+		c.metricsURL,
+	}
+
+	// Construct the request to exec into the pod
+	req := c.kubeClient.CoreV1().RESTClient().
+		Post().
+		Resource("pods").
+		Namespace(c.namespace).
+		Name(c.curlPodName).
+		SubResource("exec").
+		VersionedParams(&corev1.PodExecOptions{
+			Container: "curl",
+			Command:   cmd,
+			Stdin:     false,
+			Stdout:    true,
+			Stderr:    true,
+			TTY:       false,
+		}, scheme.ParameterCodec)
+
+	// Create an SPDY executor
+	executor, err := remotecommand.NewSPDYExecutor(c.restConfig, "POST", req.URL())
+	require.NoError(c.t, err, "Error creating SPDY executor to exec in pod")
+
+	var stdout, stderr bytes.Buffer
+	streamOpts := remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: &stdout,
+		Stderr: &stderr,
+		Tty:    false,
+	}
+
+	err = executor.StreamWithContext(ctx, streamOpts)
+	require.NoError(c.t, err, "Error streaming exec request: %v", stderr.String())
 
-	c.t.Log("Validating the metrics endpoint")
-	curlCmd := exec.Command(c.client, "exec", c.curlPodName, "-n", c.namespace, "--",
-		"curl", "-v", "-k", "-H", "Authorization: Bearer "+token, c.metricsURL)
-	output, err := curlCmd.CombinedOutput()
-	require.NoError(c.t, err, "Error calling metrics endpoint: %s", string(output))
-	require.Contains(c.t, string(output), "200 OK", "Metrics endpoint did not return 200 OK")
+	// Combine stdout + stderr
+	combined := stdout.String() + stderr.String()
+	require.Contains(c.t, combined, "200 OK", "Metrics endpoint did not return 200 OK")
```

Comment:
Is it not much more complex than before?
I think before it was very clear what it does, we could easily copy and paste to validate the things.
But anyone, maybe it is a matter of preference.

Reply:
yeah Idk, it was certainly harder to implement, `kubectl exec` was doing several things for us. I think this PR mostly just answers @joelanford concern that that help from `kubectl` also opened more avenues for things to go subtly wrong. And @azych desire that this test should use `go-client` like the other e2e.

Reply:
I could see wrapping this in a more general purpose `execInPod` kind of utility function that could be re-used. But mostly a nit since there's not another place that we need this at the moment.